### PR TITLE
Use `CryptoRandomBytes` instead of `CryptoRandomString`

### PR DESCRIPTION
**Pull request by Gusted**
- Switch to use `CryptoRandomBytes` instead of `CryptoRandomString`, OAuth's secrets are copied pasted and don't need to avoid dubious characters etc.
- `CryptoRandomBytes` gives  `CryptoRandomString` gives  possible states.
- Add a prefix, such that code scanners can easily grep these in source code.

### CODE REVIEW

File: models/auth/oauth2.go [hunk 1]
```diff
@@ -59,10 +60,12 @@ func (app *OAuth2Application) ContainsRedirectURI(redirectURI string) bool {
 
 // GenerateClientSecret will generate the client secret and returns the plaintext and saves the hash at the database
 func (app *OAuth2Application) GenerateClientSecret() (string, error) {
-	clientSecret, err := secret.New()
+	rBytes, err := util.CryptoRandomBytes(44)
```
Comment (wxiaoguang) [inline models/auth/oauth2.go]: Do we really need that long?  AES key size is only 16 or 32 bytes .....

Comment (Gusted) [reply models/auth/oauth2.go]: > Do we really need that long? AES key size is only 16 or 32 bytes .....

I've copied the length from `secret.New()`, we could reduce it to 32, that still provides enough security than the current code.

Comment (Gusted) [reply models/auth/oauth2.go]: Hmm, we might even want to add a prefix? Such that security scanners can easily grep such tokens in source codes. Github has done it as well: https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/

Comment (wxiaoguang) [reply models/auth/oauth2.go]: Then we can leave a comment in code to describe why there is a prefix 😊

**Comment by Gusted** [comment]
I've somewhere confused bits with bytes and some other things. TL;DR math was flawed, updated OP comment + code to reflect correct math.

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
`hex.EncodeToString` makes the final string length=70, it's pretty long.

Could we use base32 encoding (lower & no-padding) to reduce the size?

File: models/auth/oauth2.go [hunk 2]
```diff
@@ -394,10 +404,14 @@ func (grant *OAuth2Grant) GenerateNewAuthorizationCode(redirectURI, codeChalleng
...
 func (grant *OAuth2Grant) generateNewAuthorizationCode(e db.Engine, redirectURI, codeChallenge, codeChallengeMethod string) (code *OAuth2AuthorizationCode, err error) {
-	var codeSecret string
-	if codeSecret, err = secret.New(); err != nil {
+	rBytes, err := util.CryptoRandomBytes(34)
+	if err != nil {
 		return &OAuth2AuthorizationCode{}, err
 	}
+	// Add a prefix to the base32, this is in order to make it easier
+	// for code scanners to grab sensitive tokens.
+	codeSecret := "gta_" + base32Lower.EncodeToString(rBytes)
```
Comment (6543) [inline models/auth/oauth2.go:413]: gto_ or gta_ :D

Comment (6543) [reply models/auth/oauth2.go:413]: is there a standard/rfc ?!?

Comment (wxiaoguang) [reply models/auth/oauth2.go:413]: Free style ... https://datatracker.ietf.org/doc/html/rfc6749#section-1.4

```
Access tokens can have different formats, structures, and methods of
utilization (e.g., cryptographic properties) based on the resource
server security requirements. 
```

And https://datatracker.ietf.org/doc/html/rfc6749#appendix-A.2

```
     VSCHAR     = %x20-7E
     access-token = 1*VSCHAR
     client-secret = *VSCHAR
```


However, 34 random bytes is still very long. AES-128 is safe enough and it only uses 16 bytes keys. I prefer to shorten our keys/tokens to 16 random bytes

Comment (Gusted) [reply models/auth/oauth2.go:413]: > However, 34 random bytes is still very long. AES-128 is safe enough and it only uses 16 bytes keys. I prefer to shorten our keys/tokens to 16 random bytes

I'm sorry, but I don't understand where the AES is coming from? These are strings that should be near-impossible to guess or predict what they are. These things shouldn't be used as any form of encryption key, as these are exposed to the end-user(only once). The current strings are using 44 bytes, this PR actually already reduces them without compromising security of any sort.

Comment (Gusted) [reply models/auth/oauth2.go:413]: As for the prefix, I might hold off onto that, as that might need a better proposal/agreement on how to use that(read never in my lifetime will happen without serious motivation).

**Comment by zeripath** [comment]
CI failure is related.

---

I suspect that 34 is a bit too long - certainly it's a weird length to choose - although I know you're fitting the previous implementation if I had to guess the original length will have been chosen to fit a certain number of characters in the original encoding and to take account for the weirdness of the original algorithm. I agree that maybe we can drop back to a more sensible number of bits. 

---

I'm not sure I understand why we need these to be case-insensitive and thus why we can't use base64-no-padding here?

**Comment by Gusted** [comment]
> I'm not sure I understand why we need these to be case-insensitive and thus why we can't use base64-no-padding here?

🤷🏽 I'm a bit allergic to all the base{32,64} going around, so I'd just follow the given advice on base32.



> I suspect that 34 is a bit too long - certainly it's a weird length to choose - although I know you're fitting the previous implementation if I had to guess the original length will have been chosen to fit a certain number of characters in the original encoding and to take account for the weirdness of the original algorithm. I agree that maybe we can drop back to a more sensible number of bits.

I wanted to take 32 bits at first, but that would mean that it reduces the possible amount of states than the current implementation, which people will bark about. The only thing that I'd like with this PR is to switch to CryptoRandomBytes instead of CryptoRandomString, so any kind of encoding/bits is welcome as long it doesn't mean it will be easy to predict/brute-force it.

**Comment by wxiaoguang** [comment]
Actually, base64 doesn't help much. base64 expansion factor = log(8)/log(6) = 1.16, base32 = log(8)/log(5) = 1.29

So if we have 10 random bytes, then we have 12 base64 chars, and 13 base32 chars, not too much difference.

File: models/auth/oauth2.go (hunk 2 above)
Comment (wxiaoguang) [reply models/auth/oauth2.go:413]: >  but I don't understand where the AES is coming from.

I mean, AES-128 use 16 bytes key and is considered to be secure enough. 16 bytes is already a reasonable length for secure usage, and in many cases, AES keys are generated randomly. Although we use random bytes for random ID, it's similar. It's not wrong to use longer bytes but it is not necessary to be that long for our case.

> for the prefix

I vote for the prefixes to make the strings easier to be identified.

**Comment by Gusted** [comment]
I've now chosen for 32, which is more sensible amount of bytes, alternatives we can use 24 or 16 to lower the output length.

**Comment by 6543** [comment]
```
--- FAIL: TestOAuth2Application (4.89s)

    testlogger.go:78: 2022/01/30 12:16:16 ...eb/routing/logger.go:99:func1() [I] router: completed POST /api/v1/user/applications/oauth2 for , 201 Created in 1766.5ms @ user/app.go:196(user.CreateOauth2Application)

    api_oauth2_apps_test.go:46: 

        	Error Trace:	api_oauth2_apps_test.go:46

        	            				api_oauth2_apps_test.go:22

        	Error:      	"gto_nxmrejlw6hbco5bentvq6flqd36ts5g5lxxzmrbnwomykq7ibi2a" should have 44 item(s), but has 56

        	Test:       	TestOAuth2Application
```

Please ajust tests

**Comment by 6543** [comment]
```
--- FAIL: TestOAuth2Application (3.92s)

    api_oauth2_apps_test.go:46: 

        	Error Trace:	api_oauth2_apps_test.go:46

        	            				api_oauth2_apps_test.go:22

        	Error:      	"gto_jpioyyxqjuppuhgvq3pnqanczzp2rlx7cfiz4xhth4z5j5amabvq" should have 52 item(s), but has 56

        	Test:       	TestOAuth2Application
        	```

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
:robot:

**Pull request merged**